src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+                                   &cudnn::one,
+                                   filter_desc_,
+                                   filter_params_.Data(),
+                                   out_deriv_desc,
+                                   out_deriv.Data(),
+                                   conv_desc_,
+                                   backward_data_algo_,
+                                   work_space_,
+                                   work_space_size_,
+                                   &cudnn::one,
+                                   in_desc,
+                                   in_deriv->Data());
+  }
+
+  if (to_update) {
+    to_update->Update(in_value, out_deriv, in_desc, out_deriv_desc);
+  }
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(in_desc));
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(out_deriv_desc));
+}
+
+void CuDNN3DConvolutionComponent::Update(const CuMatrixBase<BaseFloat> &in_value,
+                                         const CuMatrixBase<BaseFloat> &out_deriv,
+                                         const cudnnTensorDescriptor_t in_desc,
+                                         const cudnnTensorDescriptor_t out_deriv_desc) {
+  cudnn::ConvolutionBackwardFilter(CuDevice::Instantiate().GetCudnnHandle(),
+                                   &learning_rate_, // alpha
+                                   in_desc,
+                                   in_value.Data(),
+                                   out_deriv_desc,
+                                   out_deriv.Data(),
+                                   conv_desc_,
+                                   backward_filter_algo_,
+                                   work_space_,
+                                   work_space_size_,
+                                   &cudnn::one, // beta
+                                   filter_desc_,
+                                   filter_params_.Data()
+                                   );
+
+  cudnn::ConvolutionBackwardBias(CuDevice::Instantiate().GetCudnnHandle(),
+                                 &learning_rate_,
+                                 out_deriv_desc,
+                                 out_deriv.Data(),
+                                 &cudnn::one,
+                                 bias_desc_,
+                                 bias_params_.Data()
+                                 );
+}
+
+Component *CuDNN3DConvolutionComponent::Copy() const {
+  CuDNN3DConvolutionComponent *ans = new CuDNN3DConvolutionComponent(*this);
+  return ans;
+}
+
+void CuDNN3DConvolutionComponent::Read(std::istream &is, bool binary) {
+  ReadUpdatableCommon(is, binary);  // Read opening tag and learning rate.
+  ExpectToken(is, binary, "<InputXDim>");
+  ReadBasicType(is, binary, &input_x_dim_);
+  ExpectToken(is, binary, "<InputYDim>");
+  ReadBasicType(is, binary, &input_y_dim_);
+  ExpectToken(is, binary, "<InputZDim>");
+  ReadBasicType(is, binary, &input_z_dim_);
```

Comment:
Add:
  ExpectToken(is, binary, "<InputNumFilters>");
  ReadBasicType(is, binary, &input_num_filters_);